src/braket/pennylane_plugin/braket_device.py

```diff
@@ -689,6 +693,107 @@ def _run_snapshots(self, snapshot_circuits, n_qubits, mapped_wires):
 
         return outcomes
 
+    def _check_pulse_frequency_validity(self, ev):
+        """Confirm that, for each waveform on the ParametrizedEvolution operator, the frequency
+        setting is a constant, and the value is within the frequency range for the relevant frame;
+        if not, raise an error"""
+
+        # confirm all frequency values are constant (or the qml.pulse.constant function)
+        callable_freqs = [
+            pulse.frequency
+            for pulse in ev.H.pulses
+            if (callable(pulse.frequency) and pulse.frequency != qml.pulse.constant)
+        ]
+
+        if callable_freqs:
+            raise RuntimeError(
+                "Expected all frequencies to be constants or qml.pulse.constant, "
+                "but recieved callable(s)"
+            )
+
+        # confirm all frequencies are within permitted difference from center frequency
+        freq_diff = (
+            self._device.properties.pulse.validationParameters["PERMITTED_FREQUENCY_DIFFERENCE"]
+            * 1e9
+        )
+        param_idx = 0
+        for pulse in ev.H.pulses:
+            freq = pulse.frequency
+            # track the index for parameters in case we need to evaluate qml.pulse.constant
+            if callable(pulse.amplitude):
+                param_idx += 1
+            if callable(pulse.phase):
+                param_idx += 1
+            if callable(pulse.frequency):
+                # if frequency is callable, its qml.pulse.constant and equal to its parameter
+                freq = ev.parameters[param_idx]
+                param_idx += 1
+
+            wires = self.map_wires(pulse.wires).tolist()
+
+            for wire in wires:
+                frame_key = f"q{wire}_drive"
+                center_freq = self._device.properties.pulse.dict()["frames"][frame_key][
+                    "centerFrequency"
+                ]
+                freq_min = center_freq - freq_diff
+                freq_max = center_freq + freq_diff
+                if not (freq_min < freq * 1e9 < freq_max):
+                    raise RuntimeError(
+                        f"Frequency range for wire {wire} is between {freq_min * 1e-9} "
+                        f"and {freq_max * 1e-9} GHz, but recieved {freq} GHz."
+                    )
+
+    def _validate_pulse_parameters(self, ev):
+        """Validates pulse input (ParametrizedEvolution) before converting to a PulseGate"""
+
+        # note: the pulse upload on the other side immediately checks that the max amplitude
```

Comment:
Yes, this means on submission of a task. We didn't add a check for max amplitude because you already get a useful error message from the service as soon as you try to create the task with a bad value for amplitude. I'll update the wording, let me know if it seems clearer.